packages/samples/src/sample-snapshot-testing.ts

```diff
@@ -0,0 +1,215 @@
+import {
+  CompilerHost,
+  NodeHost,
+  ResolveCompilerOptionsOptions,
+  compile,
+  getDirectoryPath,
+  getRelativePathFromDirectory,
+  joinPaths,
+  resolveCompilerOptions,
+  resolvePath,
+} from "@typespec/compiler";
+import { expectDiagnosticEmpty } from "@typespec/compiler/testing";
+import { fail, ok, strictEqual } from "assert";
+import { readdirSync } from "fs";
+import { mkdir, readFile, readdir, rm, writeFile } from "fs/promises";
+
+const shouldUpdateSnapshots = process.env.RECORD === "true";
+
+export interface SampleSnapshotTestOptions {
+  /**  Sample root directory. */
+  sampleDir: string;
+
+  /** Output directory for snapshots. */
+  outputDir: string;
+
+  /** Folders to exclude from testing. */
+  exclude?: string[];
+
+  /** Override the emitters to use. */
+  emit?: string[];
+}
+
+export interface TestContext {
+  runCount: number;
+  registerSnapshot(filename: string): void;
+}
+export function defineSampleSnaphotTests(config: SampleSnapshotTestOptions) {
+  const samples = resolveSamples(config);
+  let existingSnapshots: string[];
+  const writtenSnapshots: string[] = [];
+  const context = {
+    runCount: 0,
+    registerSnapshot(filename: string) {
+      writtenSnapshots.push(filename);
+    },
+  };
+  before(async () => {
+    existingSnapshots = await readFilesInDirRecursively(config.outputDir);
+  });
+
+  after(async function (this: any) {
+    if (context.runCount !== samples.length) {
+      return; // Not running the full test suite, so don't bother checking snapshots.
+    }
+
+    if (this.test.parent.tests.some((x: any) => x.state === "failed")) {
+      return; // Do not check snapshots if the test failed so we don't get a confusing error message about the missing snapshot if there is already a failure.
+    }
+
+    const missingSnapshots = new Set<string>(existingSnapshots);
```

Comment:
What if there is an additional file generated by the test run that does not show up in snapshots?